### PR TITLE
[SPARK-20962][SQL] Support subquery column aliases in FROM clause

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr added parsing rules to support subquery column aliases in FROM clause.
This pr is a sub-task of #18079.

## How was this patch tested?
Added tests in `PlanParserSuite` and `SQLQueryTestSuite`.

### CODE REVIEW

**Comment by maropu** [comment]
Jenkins, retest this please.

**Comment by maropu** [comment]
@gatorsmile If you get time, could you check this? Thanks!

**Comment by maropu** [comment]
ping

**Comment by gatorsmile** [comment]
Will review it tonight. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -753,17 +753,26 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
    * hooks.
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: We need a DDL syntax example here. Could you add it?

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: ok

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -753,17 +753,26 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
...
-    SubqueryAlias(alias, plan(ctx.queryNoWith).optionalMap(ctx.sample)(withSample))
+    val columnNamesOption = if (ctx.tableAlias.identifierList != null) {
+      Some(visitIdentifierList(ctx.tableAlias.identifierList)).map { names =>
+        names.map(n => UnresolvedAttribute(n :: Nil))
+      }
+    } else {
+      None
+    }
+
+    SubqueryAlias(alias, plan(ctx.queryNoWith).optionalMap(ctx.sample)(withSample),
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: The impact of changing `SubqueryAlias` is too big. How about adding a new operator node `UnresolvedSubqueryAlias`? In the analyzer, `UnresolvedSubqueryAlias` will be replaced by `Project` + `SubqueryAlias` ?

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: ok, I'll check if we could do so.

**Comment by maropu** [comment]
@gatorsmile ok, done. Could you check again?

**Comment by maropu** [comment]
ping

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 3]
```diff
@@ -750,20 +750,28 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
...
       // unqualified names. Users can still use this special qualifier to access columns if they
       // know it, but that's not recommended.
       "__auto_generated_subquery_name"
     } else {
-      ctx.strictIdentifier.getText
+      ctx.tableAlias.strictIdentifier.getText
+    }
+    val subquery = SubqueryAlias(alias, plan(ctx.queryNoWith).optionalMap(ctx.sample)(withSample))
+    if (ctx.tableAlias.identifierList != null) {
+      val columnNames = visitIdentifierList(ctx.tableAlias.identifierList)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Nit: `columnNames ` -> `columnAliases`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 4]
```diff
@@ -423,6 +423,26 @@ case class UnresolvedAlias(
...
+ * Aliased column names for subquery. We could add alias names for output columns in the subquery:
+ * {{{
+ *   // Assign alias names for output columns
+ *   SELECT col1, col2 FROM testData AS t(col1, col2);
+ * }}}
+ *
+ * @param outputColumnNames the column names for this subquery.
+ * @param child the logical plan of this subquery.
+ */
+case class UnresolvedSubqueryColumnAlias(
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: Nit: `UnresolvedSubqueryColumnAlias ` -> `UnresolvedSubqueryColumnAliases`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 5]
```diff
@@ -423,6 +423,26 @@ case class UnresolvedAlias(
...
 
 /**
+ * Aliased column names for subquery. We could add alias names for output columns in the subquery:
+ * {{{
+ *   // Assign alias names for output columns
+ *   SELECT col1, col2 FROM testData AS t(col1, col2);
+ * }}}
+ *
+ * @param outputColumnNames the column names for this subquery.
+ * @param child the logical plan of this subquery.
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: Nit: `the logical plan of this subquery` -> `the [[LogicalPlan]] on which this subquery column aliases apply`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 6]
```diff
@@ -423,6 +423,26 @@ case class UnresolvedAlias(
 }
 
 /**
+ * Aliased column names for subquery. We could add alias names for output columns in the subquery:
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: resolved by positions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 7]
```diff
@@ -859,6 +859,22 @@ class Analyzer(
...
       case plan if containsDeserializer(plan.expressions) => plan
 
+      case u @ UnresolvedSubqueryColumnAlias(columnNames, child) if child.resolved =>
+        // Resolves output attributes if a query has alias names in its subquery:
+        // e.g., SELECT * FROM (SELECT 1 AS a, 1 AS b) t(col1, col2)
+        val outputAttrs = child.output
+        // Checks if the number of the aliases equals to the number of output columns
+        // in the subquery.
+        if (columnNames.size != outputAttrs.size) {
+          u.failAnalysis(s"Number of column aliases does not match number of columns. " +
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Nit: remove the string Interpolator `s`.

**Comment by gatorsmile** [comment]
LGTM except a few minor comments. Thanks for working on it!

**Comment by maropu** [comment]
Thanks! Fixed.

**Comment by maropu** [comment]
@gatorsmile ok

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 8]
```diff
@@ -859,6 +859,22 @@ class Analyzer(
       // rule: ResolveDeserializer.
       case plan if containsDeserializer(plan.expressions) => plan
 
+      case u @ UnresolvedSubqueryColumnAliases(columnNames, child) if child.resolved =>
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: The last question. What is the reason why we do not have a separate analyzer rule?

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Yea, I forgot to move this rule outside. I'll update soon.

**Comment by gatorsmile** [comment]
LGTM.

Thanks! Merging to master.

**Pull request closed without merging**